### DISCUSSION TITLE
Address no/low-cost attacks submitting transactions that pass to DA

### DISCUSSION THREAD

**Discussion by mzabaluev** [movementlabsxyz/movement#475] [Ideas]
Discuss a solution for:

* #409
* #412

## Related issues

These may or may not be solved alongside.

* #416

## Proposed solutions

https://github.com/movementlabsxyz/MIP/pull/6

**Comment by l-monninger** [comment]
So, for #416, there are a couple of potentially safe quick fixes that are worth implementing.

1. Make the transactions field in our block struct a `BTreeSet` (again); I suspect I may have accidentally removed that when optimization bullets were flying. Either this or the a programmed sort has to be included for general fairness anyways.
2. Filter committed transactions from blocks before execution. 

These are discussed in a little more detail in #416. We could then send these over to @poetyellow and see what he thinks. 

To turn on #409, we have to have #416 fairly safe.

As mentioned in #416, there is still the issue of doing a fair amount of work per transaction to prevent these replay attacks thus lending to a DOS. This is where deeper conversations in https://github.com/movementlabsxyz/MIP/pull/6 and https://github.com/movementlabsxyz/MIP/pull/7 may come in. 

Without introducing some partially synchronous mechanism for the mempool, you would always have an issue with there being a period where replay transactions are inflight before their validity is agreed upon. That is we need would need to place a BFT system in front of the DA. The question then becomes, can that be less costly to operate than the DA itself?

**Reply by mzabaluev** [reply]
> Make the transactions field in our block struct a `BTreeSet` (again);

This naturally includes reviewing the `Ord` implementation on the transaction type, to make sure we can commit to this ordering and not just make an ad-hoc derived impl implicitly part of the specification.

**Reply by l-monninger** [reply]
So, do you think we would best have this be a manually implemented sort and not change the container in the type itself? This already happens AFAIK when you deserialize a `BTreeSet` structure, so performance should not be different really. 

If we do go for the sort, the average case would be that the data is (mostly) properly sorted. So, in theory, something like Insertion Sort or Timsort would be best here, I believe.

**Reply by l-monninger** [reply]
For the filter, I have not found anywhere that the `BlockExecutor` would actually be applying the same before a block is attempted to be run. So, I think this is valid:

> DB-based validation after passing through the DA is safe w.r.t. this issue, but opens up more potential for DOS. We can get the transaction by hash from the [AptosDB](https://github.com/movementlabsxyz/aptos-core/blob/7d117d7eb8052d4bc580a23094bed4e1250e1113/storage/aptosdb/src/ledger_db/transaction_db.rs#L69) it will have been called [here](https://github.com/movementlabsxyz/aptos-core/blob/7d117d7eb8052d4bc580a23094bed4e1250e1113/storage/aptosdb/src/db/include/aptosdb_writer.rs#L384) during execution if the transactions was previously included, we could filter these out of the blocks. I need to double-check if the VM is already doing this, if so, we can just use the BTreeSet to make sure it doesn't happen within a block. The problem is, however, that you're still doing a fair bit of work on every node for this transaction. Per my comment about CP earlier though, I don't think there's a good way around this. The best thing to do would probably be position the mempool for independent scaling. We are also fundamentally accepting a fair bit of [DOS risk by using a permissionless, public DA](https://forum.celestia.org/t/woods-attack-on-celestia/59) in the first place. So, some tolerance for this risk is presumed.

**Reply by l-monninger** [reply]
This is from #416 by the way.

**Reply by l-monninger** [reply]
@mzabaluev 
> I'm not sure if that addresses the issues, or if our resistance to DoS is made weaker than that provided by Celestia.

So, this does in theory add more load to every node in the network, yes. But, it shouldn't be too terribly concerning. Each one of these filter operations should be equivalent to the load of someone querying chain state over the API without the HTTP connection overhead.

The main point of the filtering, however, is to address the replay vulnerability in #416 . This foremost a risk to accounts on the network as opposed to the a DoS of the network on the whole.

**Reply by mzabaluev** [reply]
Filed #526 to investigate the aptos-executor behavior.

**Reply by mzabaluev** [reply]
I'm educated on the way aptos-executor treats this. I favor a DB-based replay check.

**Comment by l-monninger** [comment]
Just commented on #412 . I think this can also be a quick fix, but we need more information.
